clang/lib/Serialization/ASTWriterStmt.cpp

```diff
@@ -1262,6 +1262,17 @@ void ASTStmtWriter::VisitSourceLocExpr(SourceLocExpr *E) {
   Code = serialization::EXPR_SOURCE_LOC;
 }
 
+void ASTStmtWriter::VisitEmbedExpr(EmbedExpr *E) {
+  VisitExpr(E);
+  Record.AddSourceLocation(E->getBeginLoc());
+  Record.AddSourceLocation(E->getEndLoc());
+  Record.AddStmt(E->getFilenameStringLiteral());
+  Record.AddStmt(E->getDataStringLiteral());
+  Record.writeUInt32(E->getStartingElementPos());
+  Record.writeUInt32(E->getDataElementCount());
+  Code = serialization::EXPR_BUILTIN_PP_EMBED;
+}
+
```

Comment:
So, if multiple embed have the same underlying storage, and they are serialized, it duplicates the storage.
Is there a way we could avoid duplication @AaronBallman @ChuanqiXu9 ?

Reply:
I don't have context in `Preprocessor embed` and the code you quoted looks fine.

And for your question, if we have to serialize the strictly same contents, we can emit it as a record (There are many examples, one is `LocalRedeclWriter` in ASTDeclWriter::VisitRedeclarable) and remeber the offset of that record. Then when we meet it again, we can emit the offset only.

If we really need to do this, I suggest to do it in separately PRs so that it is easier to be reviewed.